core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,394 @@
+        .context("sync_block()")
+        .map_err(StorageError::Database)?
+    else {
+        return Ok(None);
+    };
+    Ok(Some(sync_block.try_into().map_err(StorageError::Database)?))
+}
+
+async fn put_block<'a>(
+    ctx: &ctx::Ctx,
+    storage: &mut StorageProcessor<'a>,
+    block: &validator::FinalBlock,
+) -> StorageResult<()> {
+    let n = MiniblockNumber(
+        block
+            .header
+            .number
+            .0
+            .try_into()
+            .context("MiniblockNumber")
+            .map_err(StorageError::Database)?,
+    );
+    let mut txn = start_transaction(ctx, storage)
+        .await
+        .wrap("start_transaction()")?;
+
+    // We require the block to be already stored in Postgres when we set the consensus field.
+    let sync_block = ctx
+        .wait(txn.sync_dal().sync_block(n, Address::default(), true))
+        .await?
+        .context("sync_block()")
+        .map_err(StorageError::Database)?
+        .context("unknown block")
+        .map_err(StorageError::Database)?;
+    let want = &ConsensusBlockFields {
+        parent: block.header.parent,
+        justification: block.justification.clone(),
+    };
+
+    // Early exit if consensus field is already set to the expected value.
+    if Some(want) == sync_block.consensus.as_ref() {
+        return Ok(());
+    }
+
+    // Verify that the payload matches the storage.
+    let payload: consensus::Payload = sync_block.try_into().map_err(StorageError::Database)?;
+    if payload.encode() != block.payload {
+        return Err(StorageError::Database(anyhow::anyhow!("payload mismatch")));
+    }
+
+    ctx.wait(txn.blocks_dal().set_miniblock_consensus_fields(n, want))
+        .await?
+        .wrap("set_miniblock_consensus_fields()")
+        .map_err(StorageError::Database)?;
+    commit(ctx, txn).await.wrap("commit()")?;
+    Ok(())
+}
+
+async fn find_head_number<'a>(
+    ctx: &ctx::Ctx,
+    storage: &mut StorageProcessor<'a>,
+) -> StorageResult<validator::BlockNumber> {
+    let head = ctx
+        .wait(
+            storage
+                .blocks_dal()
+                .get_last_miniblock_number_with_consensus_fields(),
+        )
+        .await?
+        .context("get_last_miniblock_number_with_consensus_fields()")
+        .map_err(StorageError::Database)?
+        .context("head not found")
+        .map_err(StorageError::Database)?;
+    Ok(validator::BlockNumber(head.0.into()))
+}
+
+async fn find_head_forward<'a>(
+    ctx: &ctx::Ctx,
+    storage: &mut StorageProcessor<'a>,
+    start_at: validator::BlockNumber,
+) -> StorageResult<validator::FinalBlock> {
+    let mut block = fetch_block(ctx, storage, start_at)
+        .await
+        .wrap("fetch_block()")?
+        .context("head not found")
+        .map_err(StorageError::Database)?;
+    while let Some(next) = fetch_block(ctx, storage, block.header.number.next())
+        .await
+        .wrap("fetch_block()")?
+    {
+        block = next;
+    }
```

Comment:
find_head_number takes time linear to the number of miniblocks without consensus field set, which can get quite large, especially initially when we start backfilling starting from a fixed genesis block. Note that we have a polling loop with this query, so it can get quite expensive. Sequential lookups on the other side will have amortized cost.